resolwe/flow/serializers/data.py

```diff
@@ -143,3 +143,23 @@ def get_fields(self):
             del fields['entities']
 
         return fields
+
+    def validate(self, attrs):
+        """Validate incoming data during deserialization."""
+        # Check that user has permissions on all Collections that Data
+        # object will be added to.
+        for collection_id in self.request.data.get('collections', []):
+            try:
+                collection = Collection.objects.get(pk=collection_id)
+            except Collection.DoesNotExist:
+                raise serializers.ValidationError('Invalid pk "{}" - object does not exist.'.format(collection_id))
+
+            if not self.request.user.has_perm('add_collection', obj=collection):
+                if self.request.user.has_perm('view_collection', obj=collection):
+                    msg = "You don't have `ADD` permission on collection (id: {}).".format(collection_id)
```

Comment:
It is cleaner if you insert the message directly into the exception, without an extra variable.